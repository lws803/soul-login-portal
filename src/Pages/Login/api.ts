import axios from 'axios';
import axiosRetry from 'axios-retry';

import { BASE_URL } from 'src/constants';
import { ApiError } from 'src/shared/apiTypes';

axiosRetry(axios, { retryDelay: axiosRetry.exponentialDelay, retries: 3 });

export const loginWithPlatform = async ({
  platformId,
  callback,
  values,
  state,
  codeChallenge,
}: {
  platformId: number;
  callback: string;
  values: {
    email: string;
    password: string;
  };
  state: string;
  codeChallenge: string;
}): Promise<{
  data: LoginWithPlatformResponse | null;
  error: ApiError | null;
}> => {
  try {
    const { data } = await axios.post<LoginWithPlatformResponse>(
      `${BASE_URL}/v1/auth/code`,
      values,
      {
        params: {
          client_id: platformId,
          redirect_uri: callback,
          state,
          code_challenge: codeChallenge,
        },
      },
    );
    return { data, error: null };
  } catch ({ response: { data } }) {
    return { data: null, error: data as ApiError };
  }
};

type LoginWithPlatformResponse = {
  code: string;
  state: string;
};

export const login = async ({
  email,
  password,
}: {
  email: string;
  password: string;
}) => {
  try {
    const { data } = await axios.post<LoginResponse>(
      `${BASE_URL}/v1/auth/login`,
      { email, password },
    );
    return { data, error: null };
  } catch ({ response: { data } }) {
    return { data: null, error: data as ApiError };
  }
};

type LoginResponse = {
  access_token: string;
};

export const joinPlatformAndLogin = async ({
  platformId,
  callback,
  email,
  password,
  accessToken,
  state,
  codeChallenge,
}: {
  platformId: number;
  callback: string;
  email: string;
  password: string;
  accessToken: string;
  state: string;
  codeChallenge: string;
}) => {
  try {
    await axios.post(`${BASE_URL}/v1/platforms/${platformId}/join`, undefined, {
      headers: { Authorization: `Bearer ${accessToken}` },
    });

    const { data: codeData } = await axios.post<LoginWithPlatformResponse>(
      `${BASE_URL}/v1/auth/code`,
      { email, password },
      {
        params: {
          platform_id: platformId,
          redirect_uri: callback,
          state,
          code_challenge: codeChallenge,
        },
      },
    );
    return { data: codeData, error: null };
  } catch ({ response: { data } }) {
    return { data: null, error: data as ApiError };
  }
};
