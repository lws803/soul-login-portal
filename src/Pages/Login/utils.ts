export const redirectToCallback = ({
  callback,
  code,
  state,
}: {
  callback: string;
  code: string;
  state: string;
}) => {
  if (typeof window !== 'undefined') {
    window.open(`${callback}?code=${code}&state=${state}`, '_self');
  } else {
    throw new Error("Can't redirect to callback.");
  }
};
