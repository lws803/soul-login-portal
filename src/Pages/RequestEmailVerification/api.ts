import axios from 'axios';
import axiosRetry from 'axios-retry';

import { BASE_URL } from 'src/constants';
import { ApiError } from 'src/shared/apiTypes';

axiosRetry(axios, { retryDelay: axiosRetry.exponentialDelay, retries: 3 });

export const sendEmailConfirmation = async ({ email }: { email: string }) => {
  try {
    await axios.post(
      `${BASE_URL}/v1/users/resend-confirmation-token`,
      {},
      { params: { email } },
    );
    return { error: null };
  } catch ({ response: { data } }) {
    return { error: data as ApiError };
  }
};
