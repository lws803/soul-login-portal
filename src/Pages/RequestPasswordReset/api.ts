import axios from 'axios';
import axiosRetry from 'axios-retry';

import { BASE_URL } from 'src/constants';
import { ApiError } from 'src/shared/apiTypes';

axiosRetry(axios, { retryDelay: axiosRetry.exponentialDelay, retries: 3 });

export const sendPasswordReset = async ({ email }: { email: string }) => {
  try {
    await axios.post(
      `${BASE_URL}/v1/users/request-password-reset-token`,
      {},
      { params: { email } },
    );
    return { error: null };
  } catch ({ response: { data } }) {
    return { error: data as ApiError };
  }
};
