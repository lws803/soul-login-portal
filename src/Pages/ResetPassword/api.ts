import axios from 'axios';
import axiosRetry from 'axios-retry';

import { BASE_URL } from 'src/constants';
import { ApiError } from 'src/shared/apiTypes';

axiosRetry(axios, { retryDelay: axiosRetry.exponentialDelay, retries: 3 });

export const resetPassword = async ({
  token,
  newPassword,
}: {
  token: string;
  newPassword: string;
}) => {
  try {
    await axios.post(`${BASE_URL}/v1/users/password-reset?token=${token}`, {
      password: newPassword,
    });
    return { error: null };
  } catch ({ response: { data } }) {
    return { error: data as ApiError };
  }
};
