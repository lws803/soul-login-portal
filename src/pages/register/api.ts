import axios from 'axios';

import { BASE_URL } from '../../constants';
import { ApiError } from '../../shared/api-types';

export const register = async ({
  values,
}: {
  values: { email: string; password: string; username: string };
}) => {
  try {
    const { data } = await axios.post<Response>(`${BASE_URL}/v1/users`, values);
    return { data, error: null };
  } catch ({ response: { data, status } }) {
    return {
      data: null,
      error: { data: data as ApiError, status: status as number },
    };
  }
};

type Response = {
  id?: number;
};
