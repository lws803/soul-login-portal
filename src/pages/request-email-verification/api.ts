import axios from 'axios';

import { BASE_URL } from '../../constants';
import { ApiError } from '../../shared/api-types';

export const sendPasswordReset = async ({ email }: { email: string }) => {
  try {
    await axios.post(
      `${BASE_URL}/v1/users/request-password-reset-token?email=${email}`,
    );
    return { error: null };
  } catch ({ response: { data } }) {
    return { error: data as ApiError };
  }
};
