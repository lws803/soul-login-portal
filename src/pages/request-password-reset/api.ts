import axios from 'axios';

import { BASE_URL } from '../../constants';

import { ApiError } from '../../shared/api-types';

export const sendEmailConfirmation = async ({ email }: { email: string }) => {
  try {
    await axios.post(
      `${BASE_URL}/v1/users/resend-confirmation-token?email=${email}`,
    );
    return { error: null };
  } catch ({ response: { data } }) {
    return { error: data as ApiError };
  }
};
