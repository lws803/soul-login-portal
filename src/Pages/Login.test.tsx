import React from 'react';
import { fireEvent, render, act, waitFor } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';

import Login from './Login';
import * as api from './Login/api';
import * as utils from './Login/utils';

describe(Login, () => {
  const path =
    '/?client_id=1&redirect_uri=https://www.example.com&state=STATE&code_challenge=CODE_CHALLENGE';

  it('renders', async () => {
    const { container } = render(
      <MemoryRouter initialEntries={[path]}>
        <Login />
      </MemoryRouter>,
    );

    expect(container).toMatchSnapshot();
  });

  it('renders warning when callback, platform id or state is not provided', async () => {
    const { findByText } = render(
      <MemoryRouter>
        <Login />
      </MemoryRouter>,
    );
    expect(
      await findByText(
        'Insufficient parameters provided in the url, a redirect_uri, client_id, ' +
          'state and code_challenge must be specified.',
      ),
    ).toBeInTheDocument();
  });

  it('logs in successfully', async () => {
    const loginWithPlatform = jest
      .spyOn(api, 'loginWithPlatform')
      .mockResolvedValue({
        data: { code: 'CODE', state: 'STATE' },
        error: null,
      });

    jest.spyOn(utils, 'redirectToCallback').mockImplementation();

    const { getByLabelText, getByText } = render(
      <MemoryRouter initialEntries={[path]}>
        <Login />
      </MemoryRouter>,
    );

    act(() => {
      fireEvent.change(getByLabelText('Email input'), {
        target: { value: 'TEST@EMAIL.COM' },
      });
      fireEvent.change(getByLabelText('Password input'), {
        target: { value: 'PASSWORD' },
      });
    });
    fireEvent.click(getByText('Login'));

    await waitFor(() => {
      expect(loginWithPlatform).toHaveBeenCalledWith({
        callback: 'https://www.example.com',
        platformId: 1,
        values: {
          email: 'TEST@EMAIL.COM',
          password: 'PASSWORD',
        },
        state: 'STATE',
        codeChallenge: 'CODE_CHALLENGE',
      });

      expect(utils.redirectToCallback).toHaveBeenCalledWith({
        callback: 'https://www.example.com',
        code: 'CODE',
        state: 'STATE',
      });
    });
  });
});
