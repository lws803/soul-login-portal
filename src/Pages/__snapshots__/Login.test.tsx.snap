// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login renders 1`] = `
<div>
  <div
    class="css-0"
  >
    <div
      class="css-1wiwcvy"
    >
      <div
        class="css-1ryq4o3"
      >
        <div
          class="css-1yjvs5a"
        >
          <p
            class="chakra-text css-odt1ir"
          >
            Login to your account
          </p>
          <div
            class="css-164r41r"
          >
            <p
              class="chakra-text css-0"
            >
              Not a member yet?
               
              <a
                class="chakra-link css-0"
                href="/register/?client_id=1&redirect_uri=https://www.example.com&state=STATE&code_challenge=CODE_CHALLENGE"
              >
                <strong>
                  Register now
                </strong>
              </a>
            </p>
          </div>
        </div>
        <form
          action="#"
        >
          <div
            class="chakra-form-control css-0"
            role="group"
          >
            <label
              class="chakra-form__label css-uvho5g"
              for="email"
              id="field-:r0:-label"
            >
              Email
            </label>
            <input
              aria-label="Email input"
              class="chakra-input css-0"
              id="field-:r0:"
              name="email"
              placeholder="Your email"
              value=""
            />
          </div>
          <div
            class="chakra-form-control css-164r41r"
            role="group"
          >
            <label
              class="chakra-form__label css-uvho5g"
              for="password"
              id="field-:r1:-label"
            >
              Password
            </label>
            <input
              aria-describedby="field-:r1:-helptext"
              aria-label="Password input"
              class="chakra-input css-0"
              id="field-:r1:"
              name="password"
              placeholder="Your password"
              type="password"
              value=""
            />
            <div
              class="chakra-form__helper-text css-0"
              id="field-:r1:-helptext"
            >
              <a
                class="chakra-link css-0"
                href="/request-password-reset"
                target="_blank"
              >
                Forgot password?
              </a>
            </div>
          </div>
          <div
            class="css-164r41r"
          >
            <button
              class="css-y6mgpv"
              type="submit"
            >
              <div
                class="css-tkevr6"
              >
                Login
              </div>
            </button>
          </div>
          <div
            class="css-19ug521"
          >
            <a
              class="chakra-link css-0"
              href="/request-email-verification"
              target="_blank"
            >
              Resend email verification
            </a>
          </div>
        </form>
      </div>
    </div>
    <div
      class="css-l5skqa"
    >
      <div
        class="css-18hh56p"
      >
        <div
          class="chakra-stack css-kaqv72"
        >
          <a
            class="chakra-link css-epvm6"
            href="https://api.soul-network.com/docs"
            target="_blank"
          >
            API Docs
          </a>
          <a
            class="chakra-link css-epvm6"
            href="https://status.soul-network.com"
            target="_blank"
          >
            Status
          </a>
          <a
            class="chakra-link css-epvm6"
            href="https://github.com/soul-project"
            target="_blank"
          >
            Soul's organization Github
          </a>
          <a
            class="chakra-link css-epvm6"
            href="https://discord.gg/kuGNRembKZ"
            target="_blank"
          >
            Discord server
          </a>
          <a
            class="chakra-link css-epvm6"
            href="mailto:lws803@gmail.com"
            target="_blank"
          >
            Contact me
          </a>
        </div>
        <p
          class="chakra-text css-1jtq35w"
        >
          Copyright Â© 
          2022
           Soul.
        </p>
      </div>
    </div>
  </div>
</div>
`;
